pkg/isoeditor/nmstate_handler.go

```diff
@@ -0,0 +1,134 @@
+package isoeditor
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+//go:generate mockgen -package=isoeditor -destination=mock_nmstate_handler.go . NmstateHandler
+type NmstateHandler interface {
+	CreateNmstateRamDisk(rootfsPath, ramDiskPath string) error
+}
+
+type nmstateHandler struct {
+	workDir  string
+	executer Executer
+}
+
+func NewNmstateHandler(workDir string, executer Executer) NmstateHandler {
+	return &nmstateHandler{
+		workDir:  workDir,
+		executer: executer,
+	}
+}
+
+func (n *nmstateHandler) CreateNmstateRamDisk(rootfsPath, ramDiskPath string) error {
+	// Extract nmstatectl binary
+	var err error
+	nmstateDir := filepath.Join(n.workDir, "nmstate")
+	err = os.MkdirAll(nmstateDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	binaryPath, err := n.extractNmstatectl(rootfsPath, nmstateDir)
+	if err != nil {
+		return err
+	}
+	nmstatectlPath := filepath.Join(nmstateDir, "squashfs-root", binaryPath)
+
+	// Check if nmstatectl binary file exists
+	if _, err = os.Stat(nmstatectlPath); os.IsNotExist(err) {
+		return err
+	}
+
+	// Read binary
+	nmstateBinContent, err := os.ReadFile(nmstatectlPath)
+	if err != nil {
+		return err
+	}
+
+	// Create a compressed RAM disk image with the nmstatectl binary
+	compressedCpio, err := generateCompressedCPIO(nmstateBinContent, NmstatectlPathInRamdisk, 0o100_755)
+	if err != nil {
+		return err
+	}
+
+	// Write RAM disk file
+	err = os.WriteFile(ramDiskPath, compressedCpio, 0755) //nolint:gosec
+	if err != nil {
+		return err
+	}
+
+	// Remove temp dir
+	defer func() {
+		removeErr := os.RemoveAll(nmstateDir)
+		if removeErr != nil {
+			log.WithError(removeErr).Error("failed to remove nmstate temp dir")
+		}
+	}()
+
+	return err
+}
+
+func (n *nmstateHandler) extractNmstatectl(rootfsPath, nmstateDir string) (string, error) {
+	_, err := n.executer.Execute(fmt.Sprintf("7z x %s", rootfsPath), nmstateDir)
+	if err != nil {
+		log.Errorf("failed to 7z x rootfs.img: %v", err.Error())
+		return "", err
+	}
+	// limiting files is needed on el<=9 due to https://github.com/plougher/squashfs-tools/issues/125
+	ulimit := "ulimit -n 1024"
+
+	// Listing the filesystem concisely, displaying only files (using `-lc` option).
+	// Each file in the output won't include any prefix before `/ostree` (by using `-dest ''` option),
+	// which is useful when invoking `-extract-file` (after finding the `nmstatectl` binary path).
+	list, err := n.executer.Execute(fmt.Sprintf("%s ; unsquashfs -dest '' -lc %s", ulimit, "root.squashfs"), nmstateDir)
```

Comment:
Any reason why we picked `unsquashfs` over go pkg `go-diskfs`'s squashfs?

Reply:
When I tested it, I encountered an error... the explanation according to @jhernand  is that go-diskfs does not support the zstd compression format used by CoreOS.

Reply:
I am not familiar with this area, is this a blocker or something we could get around by using go libraries for zstd in combination with go-diskfs?

Reply:
There is no workaround: squashfs compression is not something that you can plug into go-diskfs without modifying it. When I checked some time ago they had support in the master branch, but not in any release. I did test in that master branch, and it crashed. They may have improved/fixed it in the latest release.

Reply:
Thanks @jhernand for the explanation. Maybe we can leave a TODO note in the code, create a ticket for follow up and address it once the feature is mature enough? I think it's important we try to avoid shelling out processes like this if we can. WDYT?